File: Construct3/Construct3/Library/Telerik.OpenAccess.SDK/IDEIntegrations/templates/templateObjectScopeProvider.cs
//Copyright (c) %CompanyName%.  All rights reserved.
//
// usage example:
//
// // Get ObjectScope from ObjectScopeProvider
// IObjectScope scope = %ClassName%.ObjectScope();
// // start transaction
// scope.Transaction.Begin();
// // create new persistent object person and add to scope
// Person p = new Person();
// scope.Add(p);
// // commit transction
// scope.Transaction.Commit();
//

using %ProductNamespace%;
using %ProductNamespace%.Util;
//OABPE using System.Web;
//OABPE using System.Threading;

namespace %DefaultNamespace%
{
	/// <summary>
	/// This class provides an object context for connected database access.
	/// </summary>
	/// <remarks>
	/// This class can be used to obtain an IObjectScope instance required for a connected database
	/// access.
	/// </remarks>
	public class %ClassName% : IObjectScopeProvider
	{
		private Database myDatabase;
		private IObjectScope myScope;

		static private %ClassName% the%ClassName%;

		/// <summary>
		/// Constructor.
		/// </summary>
		/// <remarks></remarks>
		public %ClassName%()
		{
		}

        /// <summary>
		/// Adjusts for dynamic loading when no entry assembly is available/configurable.
		/// </summary>
		/// <remarks>
        /// When dynamic loading is used, the configuration path from the
        /// applications entry assembly to the connection setting might be broken.
        /// This method makes up the necessary configuration entries.
        /// </remarks>
        static public void AdjustForDynamicLoad()
        {
            if( the%ClassName% == null )
                the%ClassName% = new %ClassName%();

            if( the%ClassName%.myDatabase == null )
            {
                string assumedInitialConfiguration =
                           "<%RootNodeName%>" +
                               "<references>" +
                                   "<reference assemblyname='PLACEHOLDER' configrequired='True'/>" +
                               "</references>" +
                           "</%RootNodeName%>";
                System.Reflection.Assembly dll = the%ClassName%.GetType().Assembly;
                assumedInitialConfiguration = assumedInitialConfiguration.Replace(
                                                    "PLACEHOLDER", dll.GetName().Name);
                System.Xml.XmlDocument xmlDoc = new System.Xml.XmlDocument();
                xmlDoc.LoadXml(assumedInitialConfiguration);
                Database db = %ProductNamespace%.Database.Get("%ConnectionName%", 
                                            xmlDoc.DocumentElement,
                                            new System.Reflection.Assembly[] { dll } );

                the%ClassName%.myDatabase = db;
            }
        }

		/// <summary>
		/// Returns the instance of Database for the connectionId 
		/// specified in the Enable Project Wizard.
		/// </summary>
		/// <returns>Instance of Database.</returns>
		/// <remarks></remarks>
		static public Database Database()
		{
			if( the%ClassName% == null )
				the%ClassName% = new %ClassName%();

			if( the%ClassName%.myDatabase == null )
				the%ClassName%.myDatabase = %ProductNamespace%.Database.Get( "%ConnectionName%" );

			return the%ClassName%.myDatabase;
		}

		/// <summary>
		/// Returns the instance of ObjectScope for the application.
		/// </summary>
		/// <returns>Instance of IObjectScope.</returns>
		/// <remarks></remarks>
		static public IObjectScope ObjectScope()
		{
			Database();

			if( the%ClassName%.myScope == null )
				the%ClassName%.myScope = GetNewObjectScope();

			return the%ClassName%.myScope;
		}

		/// <summary>
		/// Returns the new instance of ObjectScope for the application.
		/// </summary>
		/// <returns>Instance of IObjectScope.</returns>
		/// <remarks></remarks>
		static public IObjectScope GetNewObjectScope()
		{
			Database db = Database();

			IObjectScope newScope = db.GetObjectScope(%GetObjectScopeParameters%);
			return newScope;
		}
        //OABPE /// <summary>
		//OABPE /// Returns the new instance of the ObjectScope using the HttpContext aproach described in the best practices articles.
		//OABPE /// </summary>
		//OABPE /// <returns>Instance of IObjectScope.</returns>
		//OABPE /// <remarks></remarks> 
        //OABPE public static IObjectScope GetPerRequestScope(HttpContext context) 
        //OABPE { 
        //OABPE     string key = HttpContext.Current.GetHashCode().ToString("x") + Thread.CurrentContext.ContextID.ToString(); 
        //OABPE     IObjectScope scope; 
        //OABPE     if (context == null) 
        //OABPE     { 
        //OABPE         scope = ObjectScopeProvider1.GetNewObjectScope(); 
        //OABPE     } 
        //OABPE     else 
        //OABPE     { 
        //OABPE         scope = (IObjectScope)context.Items[key]; 
        //OABPE         if (scope == null) 
        //OABPE         { 
        //OABPE             scope = ObjectScopeProvider1.GetNewObjectScope(); 
        //OABPE             context.Items[key] = scope; 
        //OABPE         } 
        //OABPE     } 
        //OABPE     return scope; 
        //OABPE }  
	}
}

File: Construct3/Construct3/Library/Telerik.OpenAccess.SDK/IDEIntegrations/templates/templateObjectScopeProvider.vb
'Copyright (c) %CompanyName%.  All rights reserved.
'
' usage example:
'
' ' Get ObjectScope from Helper
' Dim scope As IObjectScope = %ClassName%.ObjectScope
' ' start transaction
' scope.Transaction.Begin()
' ' create new persistent object person and add to scope
' Dim p As New Person
' scope.Add(p)
' ' commit transction
' scope.Transaction.Commit()
'
Imports %ProductNamespace%
Imports %ProductNamespace%.Util
'OABPE Imports System.Web
'OABPE Imports System.Threading

Public Class %ClassName%
    Implements IObjectScopeProvider
    Private _myDatabase As Database
    Private _myScope As IObjectScope

	Private Shared the%ClassName% As %ClassName%

    Public Sub New()
    End Sub

    ' Allows a dynamically loaded assembly to open the database.
    ' Must be called before the ObjectScope is aquired.
    Public Shared Sub AdjustForDynamicLoad()
        If the%ClassName% Is Nothing Then
            the%ClassName% = New %ClassName%()
        End If

        If the%ClassName%._myDatabase Is Nothing Then
            Dim assumedInitialConfiguration As String
            assumedInitialConfiguration = "<%RootNodeName%>" _
                           & "<references>" _
                           & "<reference assemblyname='PLACEHOLDER' configrequired='True'/>" _
                           & "</references>" _
                           & "</%RootNodeName%>"
            Dim repl As String = "PLACEHOLDER"
            Dim dll As System.Reflection.Assembly = the%ClassName%.GetType().Assembly
            assumedInitialConfiguration = assumedInitialConfiguration.Replace(repl, dll.GetName().Name)
            Dim xmlDoc As New System.Xml.XmlDocument
            xmlDoc.LoadXml(assumedInitialConfiguration)
            Dim allDlls As System.Reflection.Assembly() = New System.Reflection.Assembly() {dll}

            the%ClassName%._myDatabase = %ProductNamespace%.Database.Get("%ConnectionName%", _
											xmlDoc.DocumentElement, allDlls)
        End If
    End Sub

    Public Shared Function Database() As Database
		If the%ClassName% Is Nothing Then
			the%ClassName% = New %ClassName%()
        End If

		If the%ClassName%._myDatabase Is Nothing Then
			the%ClassName%._myDatabase = %ProductNamespace%.Database.Get( "%ConnectionName%" )
        End If

		Return the%ClassName%._myDatabase
    End Function

    Public Shared Function ObjectScope() As IObjectScope
        Database()

		If the%ClassName%._myScope Is Nothing Then
            the%ClassName%._myScope = GetNewObjectScope()
        End If

		Return the%ClassName%._myScope
    End Function

    Public Shared Function GetNewObjectScope() As IObjectScope
        Dim db As Database = Database()

		Dim newScope As IObjectScope = db.GetObjectScope(%GetObjectScopeParameters%)
        Return newScope
    End Function
    'OABPE Public Shared Function GetPerRequestScope(ByVal context As HttpContext) As IObjectScope
    'OABPE     Dim key As String = HttpContext.Current.GetHashCode().ToString("x") + Thread.CurrentContext.ContextID.ToString()
    'OABPE     Dim scope As IObjectScope
    'OABPE     If context Is Nothing Then
    'OABPE         scope = ObjectScopeProvider1.GetNewObjectScope()
    'OABPE     Else
    'OABPE         scope = CType(context.Items(key), IObjectScope)
    'OABPE         If scope Is Nothing Then
    'OABPE             scope = ObjectScopeProvider1.GetNewObjectScope()
    'OABPE             context.Items(key) = scope
    'OABPE         End If
    'OABPE     End If
    'OABPE     Return scope
    'OABPE End Function
End Class
